DuckDB debugging cross-post (#3)
OleMussmann/ole.mn: content/posts/2025-02-duckdb-debugging/index.md

---
date: 2025-02-15

> title: "A debugging journey into the unknown"

Reviewer: ```suggestion
title: "A Debugging Journey into the Unknown"
```
Converting to title case

summary: "We go through the steps I had to go through to debug a build problem with DuckDB, and in the process learn some concepts and techniques."
description: "An account of debugging a subtle build problem with DuckDB"
tags: ["Windows", "DLL", "DuckDB", "MingW", "cross-compilation", "git-bisect", "Julia", "BinaryBuilder"]
author: ["Suvayu Ali", "https://github.com/suvayu"]

> issueLink = "https://github.com/suvayu/suvayu/issues"

Reviewer: ```suggestion
issueLink: "https://github.com/suvayu/suvayu/issues"
```

---

![Among the Stars; Credit: NASA/JPL-Caltech](./pia17049-hires-cropped.jpg "Among the Stars; an artist's concept of Voyager in deep space")

# A debugging journey into the unknown

{{</* notice note */>}}
The story isn't complete.  There will be a concluding part when I actually fix the problem.
{{</* /notice */>}}

It all started with a colleague trying to update their Julia

> environment for one of our projects[^1].  They could not update

Reviewer: It's probably a style question, but why use footnotes instead of an inline link? Same for footnote 2 below.
```suggestion
environment for one of our projects called [Tulipa](https://github.com/TulipaEnergy).  They could not update
```

Author: I was going for:
- use link if it is relevant to the problem but not in the article because it's a lot of detail
- use footnote if it is interesting additional context, but not relevant to understand the problem

Reviewer: I see, makes sense.

I decided to dive in once and for all - this is that story.

## The issue

On Windows, trying to install any version of `DuckDB.jl` later than
`1.0.0` was failing with the error:

```julia
ERROR: LoadError: could not load symbol "duckdb_vector_size":
The specified procedure could not be found.
```

Some Julia packages, like `DuckDB.jl` depend on a native library.  The
native library is an internal dependency, and typically named
`MyPackage_jll.jl`; in the case of DuckDB, it is `DuckDB_jll.jl`.  The
error above tells us that during the compilation step, Julia tries to
load a symbol from the native library, but cannot find it.  A "symbol"
here refers to a DuckDB C-API function provided by the native DuckDB
library.

So as a first step, I wanted to check, can we actually install this
library, and load that symbol.  To get the call syntax correct, I
looked at the source code of [`DuckDB.jl`
](https://github.com/duckdb/duckdb/blob/5f5512b827df6397afd31daedb4bbdee76520019/tools/juliapkg/src/api.jl#L1285-L1297):

```julia
function duckdb_vector_size()
    return ccall((:duckdb_vector_size, libduckdb), idx_t, ())
end
```

> TThe variable `idx_t` above is defined in `ctypes.jl` as:

Reviewer: ```suggestion
The variable `idx_t` above is defined in `ctypes.jl` as:
```

```julia
const idx_t = UInt64 # DuckDB index type
```

So we can test loading the native library like this:

```julia
pkg] add DuckDB_jll
julia> using DuckDB_jll
ccall((:duckdb_vector_size, libduckdb), UInt64, ())
```

The above recipe replicates the error for any version newer than

> `1.0.0`!  Hurray!  Now that we have confirmation that the problem is

Reviewer: I think "Hooray" is more common, but I'm not sure. I follow your feeling here.

Author: US vs UK; I prefer UK

can then use other tools to inspect the [dynamic-link
libraries](https://en.wikipedia.org/wiki/Dynamic-link_library) (DLLs)
and check if the symbols actually exist.  If we can compare the
working version of the native library with a version that does not
maybe we can find out what is wrong.

On Linux, we can use `nm` from `binutils` to look at the symbols
present in the library.  Thanks to the `-C` flag, `nm` can even
"demangle"[^2] symbol names if necessary.  So for the moment we can
copy over DLLs from Windows to Linux, and inspect.

```shell
$ nm -C libduckdb.dll | grep duckdb_vector_size  # working version: v1.0.0
000000036a271d00 T duckdb_vector_size
$ nm -C libduckdb.dll | grep duckdb_vector_size  # not working version: e.g. v1.1.2
000000036a36e0e0 T duckdb_vector_size
000000036be998d0 r .rdata$.refptr.duckdb_vector_size
000000036be998d0 R .refptr.duckdb_vector_size
```

No luck :frowning:, seems that the symbols exist for both versions of
the native library.  I was puzzled.  Searching around I found while
building Windows DLLs, you have to explicitly [export symbol
names](https://learn.microsoft.com/en-us/cpp/build/exporting-from-a-dll-using-declspec-dllexport)

> using the `__declspec(dllexport)` attribute.  Besides signalling which

Reviewer: ```suggestion
using the `__declspec(dllexport)` attribute.  Besides signaling which
```
Single 'l' seems more common in both B/E and US/E.

Author: AFAIU, double-l is UK, single-l is US; I prefer UK.

load times.  So I went looking for these attributes in the DuckDB
[source
code](https://github.com/duckdb/duckdb/blob/5f5512b827df6397afd31daedb4bbdee76520019/src/include/duckdb.h#L17-L31):

```c++
#ifndef DUCKDB_API
#ifdef _WIN32
#ifdef DUCKDB_STATIC_BUILD
#define DUCKDB_API
#else
#if defined(DUCKDB_BUILD_LIBRARY) && !defined(DUCKDB_BUILD_LOADABLE_EXTENSION)
#define DUCKDB_API __declspec(dllexport)
#else
#define DUCKDB_API __declspec(dllimport)
#endif
#endif
#else
#define DUCKDB_API
#endif
#endif
```

You can see, the `#ifdef` directives conditionally defines the macro
`DUCKDB_API` which expands to `__declspec(dllexport)` when building a
Windows DLL.  Later in the header file, this macro is used to [mark
every C-API function for
export](https://github.com/duckdb/duckdb/blob/5f5512b827df6397afd31daedb4bbdee76520019/src/include/duckdb.h#L1289-L1295).

```c++
DUCKDB_API idx_t duckdb_vector_size();
```

Now that we know the C-API function name symbols are marked for export
correctly, we should check if they are indeed exported.  After some
searching, I learnt Windows development tools includes the program
`dumpbin.exe` that can show the exported symbol names.

> <details><summary>Exported symbols for the working version of <code>libduckdb.dll</code></summary>

Reviewer: ```suggestion
{{< details summary="Exported symbols for the working version of `libduckdb.dll`" >}}
```
Raw HTML is ignored for safety reasons, everything needs to be implemented via `hugo`'s shortcodes. I re-implemented `hugo`'s native details/summary shortcode implementation, so it's easier to style with CSS. The syntax stays the same, though.

        322  141 007952D0 duckdb_value_uhugeint
        323  142 007952B0 duckdb_value_uint16
        324  143 007952A0 duckdb_value_uint32
        325  144 00795290 duckdb_value_uint64
        326  145 007952C0 duckdb_value_uint8
        327  146 00792780 duckdb_value_varchar
        328  147 007957F0 duckdb_value_varchar_internal
        329  148 0078C860 duckdb_vector_assign_string_element
        330  149 0078C8C0 duckdb_vector_assign_string_element_len
        331  14A 00793090 duckdb_vector_ensure_validity_writable
        332  14B 0078C7C0 duckdb_vector_get_column_type
        333  14C 0078C810 duckdb_vector_get_data
        334  14D 0078C830 duckdb_vector_get_validity
        335  14E 0078D1B0 duckdb_vector_size

  Summary

        1000 .CRT
       12000 .bss
        F000 .data
        7000 .debug_abbrev
        1000 .debug_aranges
        3000 .debug_frame
       62000 .debug_info
        A000 .debug_line
       28000 .debug_loc
        2000 .debug_ranges
        1000 .debug_str
        3000 .edata
        5000 .idata
       81000 .pdata
      929000 .rdata
       16000 .reloc
     17E1000 .text
        1000 .tls
      17E000 .xdata
```

> </details>

Reviewer: ```suggestion
{{< /details >}}
```

> <details><summary>Exported symbols for a faulty version of <code>libduckdb.dll</code></summary>

Reviewer: ```suggestion
{{< details summary="Exported symbols for a faulty version of `libduckdb.dll`" >}}
```

         74   49 01500AE0 _ZN6duckdb17NumericValueUnion18GetReferenceUnsafeIxEERT_v
         75   4A 01500AF0 _ZN6duckdb17NumericValueUnion18GetReferenceUnsafeIyEERT_v
         76   4B 0177C5D0 _ZN6duckdb7Catalog17GetCatalogVersionERNS_13ClientContextE
         77   4C 0177C5E0 _ZN6duckdb7Catalog29CheckAmbiguousCatalogOrSchemaERNS_13ClientContextERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
         78   4D 018BFE90 _ZNK6duckdb7Catalog21CatalogTypeLookupRuleENS_11CatalogTypeE
         79   4E 018BFFE0 _ZNK6duckdb9ErrorData10RawMessageB5cxx11Ev
         80   4F 018BFFF0 _ZNK6duckdb9ErrorData7MessageB5cxx11Ev
         81   50 018C0000 _ZNK6duckdb9Extension7VersionB5cxx11Ev
         82   51 00464E50 duckdb_adbc_init

  Summary

        1000 .CRT
       12000 .bss
        F000 .data
        8000 .debug_abbrev
        1000 .debug_aranges
        3000 .debug_frame
       63000 .debug_info
        A000 .debug_line
       28000 .debug_loc
        2000 .debug_ranges
        1000 .debug_str
        1000 .edata
        5000 .idata
       8A000 .pdata
      9C2000 .rdata
       17000 .reloc
     19EB000 .text
        1000 .tls
      197000 .xdata
```

> </details>

Reviewer: ```suggestion
{{< /details >}}
```

set(CMAKE_RC_COMPILER ${TOOLCHAIN_PREFIX}-windres)

# target environment on the build host system
set(CMAKE_FIND_ROOT_PATH /usr/${TOOLCHAIN_PREFIX})

# modify default behavior of FIND_XXX() commands
set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
```

Now we can use this script with `git-bisect` to find the first bad
commit, like this:

```bash
$ git bisect start
$ git bisect bad v1.1.2
$ git bisect good v1.0.0
$ git bisect run ./bisect-script.bash
$ git bisect visualize # shows a nice summary
$ git bisect reset

```

The bisection led me to this commit:
[d1ea1538](https://github.com/duckdb/duckdb/commit/d1ea1538c9217fb536485f1500f04a0b55b1e584).
Unfortunately it is not clear how that commit would lead to symbol
export failure, it only adds 11 new functions to the C-API (so 11 new
symbols).  So for now, this debugging journey has to stop here,
without a clear resolution.  But we did learn a lot of new concepts,
and used a wide variety of tools to investigate.

## What did we learn?

To summarise, the issue is:
1. All releases of `DuckDB.jl` after `v1.0.0` on Windows are broken.

> 2. We recreate the issue without Julia, which shows the the DuckDB

Reviewer: ```suggestion
2. We recreate the issue without Julia, which shows the DuckDB
```
